File: study/lua/src/app/scene/view/homeland/PopupHomelandSubUp.lua

-- Author: hedili
-- Date:2018-05-02 15:59:38
-- Describle： 子树升级界面

local PopupBase = require("app.ui.PopupBase")
local PopupHomelandSubUp = class("PopupHomelandSubUp", PopupBase)
local HomelandConst = require("app.const.HomelandConst")
local HomelandHelp = require("app.scene.view.homeland.HomelandHelp")
local PopupHomelandUpSubCell = require("app.scene.view.homeland.PopupHomelandUpSubCell")

local UserDataHelper = require("app.utils.UserDataHelper")
local TextHelper = require("app.utils.TextHelper")

function PopupHomelandSubUp:ctor( treeData, isFriend)
	self._treeData = treeData
	self._isFriend = isFriend
	self._attrNode1 = nil  --SingleNode
	self._attrNode2 = nil  --SingleNode
	self._attrNode3 = nil  --SingleNode
	self._btnUp = nil  --CommonButtonHighLight
	self._commonHelp = nil  --CommonHelp
	self._commonNodeBk = nil  --CommonNormalSmallPop
	self._conditionTitle = nil  --ImageView
	self._imageArrow = nil  --ImageView
	self._resItem1 = nil  --CommonResourceInfo
	self._resItem2 = nil  --CommonResourceInfo
	self._resItem3 = nil  --CommonResourceInfo
	self._textAddPlayerJoint = nil  --Text
	self._textNodeRes = nil  --SingleNode
	self._textRankMax = nil  --Text

	local resource = {
		file = Path.getCSB("PopupHomelandSubUp", "homeland"),
		binding = {
			_btnUp = {
				events = {{event = "touch", method = "_onBtnSubUp"}}
			},
		},
	}

	self:setName("PopupHomelandSubUp")

	PopupHomelandSubUp.super.ctor(self, resource)
end

-- Describle：
function PopupHomelandSubUp:onCreate()
	self._commonNodeBk:setTitle(Lang.get("homeland_popup_title"))
	self._commonNodeBk:addCloseEventListener(handler(self, self.onBtnCancel))
	self._btnUp:setString(Lang.get("homeland_lv_up_btn"))
	self:_createDlgCell()

	self._nodeMax:setVisible(false)
	self._nodeSubTree:setVisible(false)

	-- self:_updateUI()

	if self._isFriend == true then
		self:showFriend()
	end
end

function PopupHomelandSubUp:_createDlgCell( ... )
	-- body

	self._levelUpCell1 = PopupHomelandUpSubCell.new()
	self._attrNode1:addChild(self._levelUpCell1)

	self._levelUpCell3 = PopupHomelandUpSubCell.new()
	self._attrNode3:addChild(self._levelUpCell3)
	
	self._levelUpCell3:setVisible(false)
	self._textRankMax:setVisible(false)
end

function PopupHomelandSubUp:onBtnCancel()
	if not isBreak then
		self:close()
	end
end

function PopupHomelandSubUp:_getSubCfg( ... )
	-- body
	local subData1, subData2 = HomelandHelp.getSubTreeCfg(self._treeData)
	dump(subData1)
	dump(subData2)
	return subData1, subData2
end

function PopupHomelandSubUp:_updateUI( ... )
	-- body
	logWarn("PopupHomelandSubUp:_updateUI")

	local subData1, subData2 = self:_getSubCfg()
	local isMax = self:updateSubTree()
	if isMax == false then
		self:_updateSubCostInfo(subData1,subData2)
	end

end

--限制装饰类型
function PopupHomelandSubUp:_updateSubTreeCondition( subData1,subData2 )
	-- body
	local limitList= {}

	local rootNode = self._nodeSubTree:getSubNodeByName("FileNode_level3")
	rootNode:setVisible(false)
	if subData2.limit_tree_level > 0 then
		table.insert(limitList, {type = 0, level = subData2.limit_tree_level})
	end
	

	for i = 1 , 2 do
		local rootNode = self._nodeSubTree:getSubNodeByName("FileNode_level"..i)
		rootNode:setVisible(false)
		local subType = subData2["adorn_type_"..i]
		local subLevel = subData2["adorn_level_"..i]
		if subType and subType > 0 and subLevel and subLevel > 0 then
			table.insert(limitList, {type = subType, level = subLevel})
		end
	end
	local function updateSubTreeLimit(rootNode, value )
		-- body
		local descName = ""
		local levelOk = false
		if value.type == 0 then
			local mainTree = G_UserData:getHomeland():getMainTreeCfg(value.level)
			descName = mainTree.name.." : "
			local mainTreeLevel = G_UserData:getHomeland():getMainTreeLevel()
			levelOk = mainTreeLevel >= value.level 
		else
			local subCfg = G_UserData:getHomeland():getSubTreeCfg(value.type, value.level)
			local subLevel = G_UserData:getHomeland():getSubTreeLevel(value.type)
			levelOk = subLevel >= value.level 
			descName = subCfg.name.." :"
		end

		
		rootNode:setVisible(true)
		local textName = rootNode:getSubNodeByName("Text_name")
		local textLevel = rootNode:getSubNodeByName("Text_value")
		local button = rootNode:getSubNodeByName("Button_go_level")
		local desc = Lang.get("homeland_sub_tree_desc", {num= value.level} )
		if value.type == 0 then
			desc = Lang.get("homeland_main_tree_desc", {num = value.level})
		end
		
		local color = Colors.DARK_BG_GREEN
		local function goSubTree( ... )
			-- body
			G_SignalManager:dispatch(SignalConst.EVENT_HOME_LAND_OPEN_DLG, value.type)
			self:close()
			logWarn("goSubTree   xcxxxxxxxxxxxxxxx")
		end
		button:setVisible(not levelOk)
		if levelOk == false then
			if value.type == 0 then
				desc = Lang.get("homeland_main_tree_desc_no", {num= value.level} )
			else
				desc = Lang.get("homeland_sub_tree_desc_no", {num= value.level} )
			end
			
			color = Colors.DARK_BG_RED
			button:addClickEventListenerEx(goSubTree)
		end
		textName:setString(descName)
		textLevel:setString(desc)
		textLevel:setColor(color)
	end

	for i, value in ipairs(limitList) do
		local rootNode = self._nodeSubTree:getSubNodeByName("FileNode_level"..i)

		updateSubTreeLimit(rootNode, value )
	end

	
end


function PopupHomelandSubUp:_updateSubCostInfo( subData1,subData2)
	-- body
	local costList = {}
	for i=1, 2 do
		local type = subData1["type_"..i]
		local value = subData1["value_"..i]
		local size = subData1["size_"..i]
		if type > 0 and value > 0 then
			table.insert(costList, {type = type, value =value, size = size})
		end
		self["_resSubItem"..i]:setVisible(false)
	end

	for i, value in ipairs(costList) do
		local costNode = self["_resSubItem"..i]
		costNode:updateView(value)
		costNode:setVisible(true)
	end
	self:_updateSubTreeCondition(subData1,subData2)
end





function PopupHomelandSubUp:_showMax( ... )
	-- body
	self._nodeMaxAttr:setVisible(false)
	self._textRankMax:setVisible(false)
	self._nodeMax:setVisible(true)
	local subData1, subData2 = self:_getSubCfg()
	self._levelUpCell1:setVisible(false)
	self._levelUpCell3:setVisible(true)
	self._levelUpCell3:updateUI(subData1, subData2)
	
	if subData2 == nil then
		self._textRankMax:setVisible(true)
		self._levelUpCell3:moveAttrToMid()
	else
		self._nodeMaxAttr:setVisible(true)
	end

end

--好友访问
function PopupHomelandSubUp:showFriend( ... )
	-- body
	self:_showMax()
	self._textRankMax:setVisible(false)
	self._nodeSubTree:setVisible(false)
end


function PopupHomelandSubUp:updateSubTree( )
	-- body
	logWarn("PopupHomelandSubUp:updateSubTree")

	local subData1, subData2 = self:_getSubCfg()
	--max
	self._nodeMax:setVisible(false)
	self._nodeSubTree:setVisible(false)

	if subData2 == nil then
		self:_showMax()
		return true
	end

	--two
	if subData1 and subData2 then
		self._nodeSubTree:setVisible(true)
		self._levelUpCell1:setVisible(true)

		self._levelUpCell3:setVisible(false)
		self._levelUpCell1:updateUI(subData1, subData2)
	end

	return false
end

-- Describle：
function PopupHomelandSubUp:onEnter()


	self._signalHomeTreeUpLevel = G_SignalManager:add(SignalConst.EVENT_HOME_TREE_UP_LEVEL_SUCCESS, handler(self, self._onEventHomeTreeUpLevel))


	self._signalRecvCurrencysInfo = G_SignalManager:add(SignalConst.EVENT_RECV_CURRENCYS_INFO, handler(self, self._updateData))
	
	self:_updateUI()
end

function PopupHomelandSubUp:onShowFinish( ... )
	-- body
	--抛出新手事件出新手事件
    G_SignalManager:dispatch(SignalConst.EVENT_TUTORIAL_STEP, self.__cname)
end

-- Describle：
function PopupHomelandSubUp:onExit()
	self._signalHomeTreeUpLevel:remove()
	self._signalHomeTreeUpLevel = nil

	self._signalRecvCurrencysInfo:remove()
	self._signalRecvCurrencysInfo = nil
end

-- Describle：
function PopupHomelandSubUp:_onBtnSubUp( ... )
	-- body
	local retValue = HomelandHelp.checkSubTreeUp( self._treeData )
	if retValue then
		--id 为0 升级主树
		G_UserData:getHomeland():c2sHomeTreeUpLevel( self._treeData.treeId )
	end
end


function PopupHomelandSubUp:_onEventHomeTreeUpLevel(id, message)	
	--G_Prompt:showTip("晋级成功")
	if self._isFriend == true then
		return
	end
	self:close()
end


function PopupHomelandSubUp:_updateData()

	if self._isFriend == true then
		return
	end
	self:_updateUI()
end

return PopupHomelandSubUp

File: study/ts/script/scene/view/homeland/PopupHomelandSubUp.ts
const { ccclass, property } = cc._decorator;

import CommonCostNode from '../../../ui/component/CommonCostNode'

import CommonButtonLevel0Highlight from '../../../ui/component/CommonButtonLevel0Highlight'

import CommonNormalLargePop from '../../../ui/component/CommonNormalLargePop'
import PopupBase from '../../../ui/PopupBase';
import { Lang } from '../../../lang/Lang';
import { handler } from '../../../utils/handler';
import PopupHomelandUpSubCell from './PopupHomelandUpSubCell';
import { HomelandHelp } from './HomelandHelp';
import { G_UserData, Colors, G_SignalManager } from '../../../init';
import { SignalConst } from '../../../const/SignalConst';
import UIHelper from '../../../utils/UIHelper';

@ccclass
export default class PopupHomelandSubUp extends PopupBase {
    public static path = 'homeland/PopupHomelandSubUp';
    @property({
        type: CommonNormalLargePop,
        visible: true
    })
    _commonNodeBk: CommonNormalLargePop = null;

    @property({
        type: cc.Node,
        visible: true
    })
    _nodeSubTree: cc.Node = null;

    @property({
        type: cc.Node,
        visible: true
    })
    _textNodeRes: cc.Node = null;

    @property({
        type: CommonButtonLevel0Highlight,
        visible: true
    })
    _btnUp: CommonButtonLevel0Highlight = null;

    @property({
        type: CommonCostNode,
        visible: true
    })
    _resSubItem1: CommonCostNode = null;

    @property({
        type: CommonCostNode,
        visible: true
    })
    _resSubItem2: CommonCostNode = null;

    @property({
        type: cc.Node,
        visible: true
    })
    _nodeMax: cc.Node = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textRankMax: cc.Label = null;

    @property({
        type: cc.Node,
        visible: true
    })
    _nodeMaxAttr: cc.Node = null;

    @property({
        type: cc.Node,
        visible: true
    })
    _attrNode1: cc.Node = null;

    @property({
        type: cc.Node,
        visible: true
    })
    _attrNode3: cc.Node = null;
    @property({
        type: cc.Prefab,
        visible: true
    })
    popupHomelandUpSubCell: cc.Prefab = null;

    _treeData: any;
    _isFriend: any;
    _signalHomeTreeUpLevel;
    _signalRecvCurrencysInfo;
    _levelUpCell1: PopupHomelandUpSubCell;
    _levelUpCell3: PopupHomelandUpSubCell;
    ctor(treeData, isFriend) {
        this._treeData = treeData;
        this._isFriend = isFriend;
        this._isClickOtherClose = true;
    }
    onCreate() {
        this.node.name = "PopupHomelandSubUp";
        this._btnUp.addClickEventListenerEx(handler(this, this._onBtnSubUp));
        this._commonNodeBk.setTitle(Lang.get('homeland_popup_title'));
        this._commonNodeBk.addCloseEventListener(handler(this, this.onBtnCancel));
        this._btnUp.setString(Lang.get('homeland_lv_up_btn'));
        this._createDlgCell();
        this._nodeMax.active = (false);
        this._nodeSubTree.active = (false);
        if (this._isFriend == true) {
            this.showFriend();
        }
    }
    _createDlgCell() {
        this._levelUpCell1 = cc.instantiate(this.popupHomelandUpSubCell).getComponent(PopupHomelandUpSubCell);
        this._attrNode1.addChild(this._levelUpCell1.node);
        this._levelUpCell3 = cc.instantiate(this.popupHomelandUpSubCell).getComponent(PopupHomelandUpSubCell);
        this._attrNode3.addChild(this._levelUpCell3.node);
        this._levelUpCell3.node.active = (false);
        this._textRankMax.node.active = (false);
    }
    onBtnCancel() {
        this.close();
    }
    _getSubCfg() {
        var [subData1, subData2] = HomelandHelp.getSubTreeCfg(this._treeData);
        return [
            subData1,
            subData2
        ];
    }
    _updateUI() {
        var [subData1, subData2] = this._getSubCfg();
        var isMax = this.updateSubTree();
        if (isMax == false) {
            this._updateSubCostInfo(subData1, subData2);
        }
    }
    _updateSubTreeCondition(subData1, subData2) {
        var limitList = [];
        var rootNode = this._nodeSubTree.getChildByName('FileNode_level3');
        rootNode.active = (false);
        if (subData2.limit_tree_level > 0) {
            limitList.push({
                type: 0,
                level: subData2.limit_tree_level
            });
        }
        for (var i = 1; i <= 2; i++) {
            var rootNode = this._nodeSubTree.getChildByName('FileNode_level' + i);
            rootNode.active = (false);
            var subType = subData2['adorn_type_' + i];
            var subLevel = subData2['adorn_level_' + i];
            if (subType && subType > 0 && subLevel && subLevel > 0) {
                limitList.push({
                    type: subType,
                    level: subLevel
                });
            }
        }
        var  updateSubTreeLimit = function(rootNode, value) {
            var descName = '';
            var levelOk = false;
            if (value.type == 0) {
                var mainTree = G_UserData.getHomeland().getMainTreeCfg(value.level);
                descName = mainTree.name + ' : ';
                var mainTreeLevel = G_UserData.getHomeland().getMainTreeLevel();
                levelOk = mainTreeLevel >= value.level;
            } else {
                var subCfg = G_UserData.getHomeland().getSubTreeCfg(value.type, value.level);
                var subLevel = G_UserData.getHomeland().getSubTreeLevel(value.type);
                levelOk = subLevel >= value.level;
                descName = subCfg.name + ' :';
            }
            rootNode.active = (true);
            var node = rootNode.getChildByName('Panel_attr');
            var textName = node.getChildByName('Text_name');
            var textLevel = node.getChildByName('Text_value');
            var button = node.getChildByName('Button_go_level');
            var desc = Lang.get('homeland_sub_tree_desc', { num: value.level });
            if (value.type == 0) {
                desc = Lang.get('homeland_main_tree_desc', { num: value.level });
            }
            var color = Colors.DARK_BG_GREEN;
            button.active = (!levelOk);
            if (levelOk == false) {
                if (value.type == 0) {
                    desc = Lang.get('homeland_main_tree_desc_no', { num: value.level });
                } else {
                    desc = Lang.get('homeland_sub_tree_desc_no', { num: value.level });
                }
                color = Colors.DARK_BG_RED;
                UIHelper.addEventListener(this.node, button.getComponent(cc.Button) || button.addComponent(cc.Button), 'PopupHomelandSubUp', 'goSubTree', value);
            }
            textName.getComponent(cc.Label).string = (descName);
            textLevel.getComponent(cc.Label).string =(desc);
            textLevel.color = (color);
        }.bind(this);
        for (var j in limitList) {
            var value = limitList[j];
            var rootNode = this._nodeSubTree.getChildByName('FileNode_level' + (parseFloat(j) + 1));
            updateSubTreeLimit(rootNode, value);
        }
    }

    goSubTree(evetn, value) {
        G_SignalManager.dispatch(SignalConst.EVENT_HOME_LAND_OPEN_DLG, value.type);
        this.close();
    }
    _updateSubCostInfo(subData1, subData2) {
        var costList = [];
        for (var i = 1; i <= 2; i++) {
            var type = subData1['type_' + i];
            var value = subData1['value_' + i];
            var size = subData1['size_' + i];
            if (type > 0 && value > 0) {
                costList.push({
                    type: type,
                    value: value,
                    size: size
                });
            }
            this['_resSubItem' + i].node.active = (false);
        }
        for (var j in costList) {
            var value = costList[j];
            var costNode = this['_resSubItem' + (parseFloat(j) + 1)];
            costNode.updateView(value);
            costNode.node.active = (true);
        }
        this._updateSubTreeCondition(subData1, subData2);
    }
    _showMax() {
        this._nodeMaxAttr.active = (false);
        this._textRankMax.node.active = (false);
        this._nodeMax.active = (true);
        var [subData1, subData2] = this._getSubCfg();
        this._levelUpCell1.node.active = (false);
        this._levelUpCell3.node.active = (true);
        this._levelUpCell3.updateUI(subData1, subData2);
        if (subData2 == null) {
            this._textRankMax.node.active = (true);
            this._levelUpCell3.moveAttrToMid();
        } else {
            this._nodeMaxAttr.active = (true);
        }
    }
    showFriend() {
        this._showMax();
        this._textRankMax.node.active = (false);
        this._nodeSubTree.active = (false);
    }
    updateSubTree() {
        var [subData1, subData2] = this._getSubCfg();
        this._nodeMax.active = (false);
        this._nodeSubTree.active = (false);
        if (subData2 == null) {
            this._showMax();
            return true;
        }
        if (subData1 && subData2) {
            this._nodeSubTree.active = (true);
            this._levelUpCell1.node.active = (true);
            this._levelUpCell3.node.active = (false);
            this._levelUpCell1.updateUI(subData1, subData2);
        }
        return false;
    }
    onEnter() {
        this._signalHomeTreeUpLevel = G_SignalManager.add(SignalConst.EVENT_HOME_TREE_UP_LEVEL_SUCCESS, handler(this, this._onEventHomeTreeUpLevel));
        this._signalRecvCurrencysInfo = G_SignalManager.add(SignalConst.EVENT_RECV_CURRENCYS_INFO, handler(this, this._updateData));
        this._updateUI();
    }
    onShowFinish() {
        G_SignalManager.dispatch(SignalConst.EVENT_TUTORIAL_STEP, this.node.name);
    }
    onExit() {
        this._signalHomeTreeUpLevel.remove();
        this._signalHomeTreeUpLevel = null;
        this._signalRecvCurrencysInfo.remove();
        this._signalRecvCurrencysInfo = null;
    }
    _onBtnSubUp() {
        var retValue = HomelandHelp.checkSubTreeUp(this._treeData);
        if (retValue) {
            G_UserData.getHomeland().c2sHomeTreeUpLevel(this._treeData.treeId);
        }
    }
    _onEventHomeTreeUpLevel(id, message) {
        if (this._isFriend == true) {
            return;
        }
        this.close();
    }
    _updateData() {
        if (this._isFriend == true) {
            return;
        }
        this._updateUI();
    }

}